django/db/models/expressions.py

```diff
@@ -866,6 +895,112 @@ def as_sql(self, compiler, connection, template=None, case_joiner=None, **extra_
+        clone.subquery.query.bump_prefix(query)
+
+        # Need to recursively resolve these.
+        def resolve_all(child):
+            if hasattr(child, 'children'):
+                [resolve_all(_child) for _child in child.children]
+            if hasattr(child, 'rhs'):
+                child.rhs = resolve(child.rhs)
+
+        def resolve(child):
+            if hasattr(child, 'resolve_expression'):
+                return child.resolve_expression(
+                    query=query, allow_joins=allow_joins, reuse=reuse, summarize=summarize,
+                    for_save=for_save)
+            return child
+
+        resolve_all(clone.subquery.query.where)
+
+        for key, value in clone.subquery.query.annotations.items():
+            if isinstance(value, SubQuery):
+                clone.subquery.query.annotations[key] = resolve(value)
+
+        return clone
+
+    def get_source_expressions(self):
+        return [x for x in [getattr(expr, 'lhs', None) for expr in self.subquery.query.where.children] if x]
+
+    def relabeled_clone(self, change_map):
+        clone = self.copy()
+        clone.subquery.query = clone.subquery.query.relabeled_clone(change_map)
+        clone.subquery.query.external_aliases.update(
+            alias for alias in change_map.values() if alias not in clone.subquery.query.tables)
+        return clone
+
+    def as_sql(self, compiler, connection, template=None, **extra_context):
+        connection.ops.check_expression_support(self)
+        template_params = self.extra.copy()
+        template_params.update(extra_context)
+
+        template_params['subquery'], sql_params = self.subquery.query.get_compiler(connection=connection).as_sql()
+
+        template = template or template_params.get('template', self.template)
+        sql = template % template_params
+        sql = connection.ops.unification_cast_sql(self.output_field) % sql
+        return sql, sql_params
+
+    def _prepare(self, output_field):
+        # If we are the rhs in a subquery, we want to remove the wrapping ()
+        if self.template == '(%(subquery)s)':
+            clone = self.copy()
+            clone.template = '%(subquery)s'
+            return clone
+
+        return self
+
+
+class Exists(SubQuery):
+    template = 'EXISTS(%(subquery)s)'
+
+    def __init__(self, *args, **kwargs):
+        self.negated = kwargs.pop('negated', False)
+        super(Exists, self).__init__(*args, **kwargs)
+
+    def __invert__(self):
+        return type(self)(self.subquery, self.output_field, negated=(not self.negated), **self.extra)
+
+    @property
+    def output_field(self):
+        return fields.BooleanField()
+
+    def resolve_expression(self, query=None, **kwargs):
+        # By definition, an EXISTS does not care about the columns, so the query can
+        # be simplified down to the primary key.
+        self.subquery = self.subquery.values('pk').order_by()
+        return super(Exists, self).resolve_expression(query, **kwargs)
```

Comment:
The comment above is only partially true. For aggregations within the EXISTS, if you select the primary key, then you force the primary key into the GROUP BY and evict the user specified grouping. I *think* it'd be more correct to make the user select a single column in their subquery. Again, here's a trivial example:

```python
inner = (
    FTest.objects
         .filter(talk=OuterRef('num_int'))
         .values('talk')
         .annotate(talk_sum=Sum('talk'))
)

qs = Constraints.objects.annotate(match=Exists(inner)).filter(match=True)
```

```sql
SELECT "scratch_constraints"."id",
"scratch_constraints"."num_int",
"scratch_constraints"."num_float",
"scratch_constraints"."num_dec",
"scratch_constraints"."char",
"scratch_constraints"."text",
EXISTS(
    SELECT U0."id" 
    FROM "scratch_ftest" U0 
    WHERE U0."talk" = ("scratch_constraints"."num_int") 
    GROUP BY U0."id"
) AS "match" FROM "scratch_constraints" 
WHERE 
    EXISTS(
        SELECT U0."id" 
        FROM "scratch_ftest" U0 
        WHERE U0."talk" = ("scratch_constraints"."num_int") 
        GROUP BY U0."id"
    ) = True
```

I'm also not sure whether the EXISTS in the WHERE clause needs to be written like that, or if it should just reference the alias of the EXISTS column. That is, I'm unsure if using the alias in the WHERE clause is valid SQL for all dialects we support. I don't think this is critical to solve in a first pass though.